geoviews/operation/resample.py

```diff
@@ -0,0 +1,197 @@
+    into.
+
+    Parameters
+    ----------
+    bounds: shapely.geometry.Polygon
+        Polygon representing the area of the current viewport
+    domain: shapely.geometry.Polygon
+        Polygon representing the overall bounding region of the data
+    levels: int
+        Number of zoom levels to divide the domain into
+
+    Returns
+    -------
+    zoom_level: int
+        Integer zoom level
+    """
+    area_fraction = min(bounds.area/domain.area, 1)
+    return int(min(round(np.log2(1/area_fraction)), levels))
+
+
+def bounds_to_poly(bounds):
+    """
+    Constructs a shapely Polygon from the provided bounds tuple.
+
+    Parameters
+    ----------
+    bounds: tuple
+        Tuple representing the (left, bottom, right, top) coordinates
+
+    Returns
+    -------
+    polygon: shapely.geometry.Polygon
+        Shapely Polygon geometry of the bounds
+    """
+    x0, y0, x1, y1 = bounds
+    return Polygon([(x0, y0), (x1, y0), (x1, y1), (x0, y1)])
+
+
+class resample_geometry(Operation):
+    """
+    This operation dynamically culls and resamples Path or Polygons
+    elements based on the current zoom level. On first execution a
+    RTree created using the Sort-Tile-Recursive algorithm, which is
+    used to query for geometries within the current viewport (defined
+    by the x_range and y_range).
+
+    Any geometries returned by the RTree query are tested to ensure
+    their area is over the display_threshold, expressed as a fraction
+    of the current viewport area. Any remaining polygons are
+    simplified using the Douglas-Peucker algorithm, which eliminates
+    vertices while ensuring that the curve does not diverge from the
+    original curve by more than the tolerance. The tolerance is
+    expressed as a fraction of the square root of the area of the
+    current viewport.
+
+    Once computed a simplified geometry is cached depending on the
+    current zoom level. The number of valid zoom levels can be
+    declared and are used to recursively subdivide the domain into
+    smaller subregions.
+
+    If requested the geometries can also be clipped to the current
+    viewport which avoids having to render vertices that are not
+    visible.
+    """
+
+    cache = param.Boolean(default=True, doc="""
+        Whether to cache simplified geometries depending on the zoom
+        level.""")
+
+    clip = param.Boolean(default=False, doc="""
+        Whether to disable the cache and clip polygons
+        to current bounds.""")
+
+    display_threshold = param.Number(default=0.0001, doc="""
+        The fraction of the current viewport covered by a geometry
+        before it is shown.""")
+
+    dynamic = param.Boolean(default=True, doc="""
+       Enables dynamic processing by default.""")
+
+    preserve_topology = param.Boolean(default=False, doc="""
+        Whether to preserve topology between geometries
+        (much faster when disabled).""")
```

Comment:
Can you explain what the tradeoff is?  What does it mean to preserve topology?

Reply:
I'll update the docstring, basically it avoids self-intersecting or otherwise invalid geometries but is **much** slower.